receiver/googlecloudspannerreceiver/internal/statsreader/intervalstatsreader.go

```diff
@@ -0,0 +1,149 @@
+	lastPullTimestamp time.Time
+}
+
+func newIntervalStatsReaderWithMaxRowsLimit(
+	logger *zap.Logger,
+	database *datasource.Database,
+	metricsMetadata *metadata.MetricsMetadata,
+	backFillEnabled bool,
+	topMetricsQueryMaxRows int) *intervalStatsReader {
+
+	reader := currentStatsReader{
+		logger:                 logger,
+		database:               database,
+		metricsMetadata:        metricsMetadata,
+		topMetricsQueryMaxRows: topMetricsQueryMaxRows,
+		statement:              intervalStatsStatement,
+	}
+
+	return &intervalStatsReader{
+		currentStatsReader: reader,
+		backFillEnabled:    backFillEnabled,
+	}
+}
+
+func newIntervalStatsReader(
+	logger *zap.Logger,
+	database *datasource.Database,
+	metricsMetadata *metadata.MetricsMetadata,
+	backFillEnabled bool) *intervalStatsReader {
+
+	reader := currentStatsReader{
+		logger:          logger,
+		database:        database,
+		metricsMetadata: metricsMetadata,
+		statement:       intervalStatsStatement,
+	}
+
+	return &intervalStatsReader{
+		currentStatsReader: reader,
+		backFillEnabled:    backFillEnabled,
+	}
+}
+
+func (reader *intervalStatsReader) Read(ctx context.Context) ([]pdata.Metrics, error) {
+	reader.logger.Info(fmt.Sprintf("Executing read method for reader %v", reader.Name()))
+
+	// Generating pull timestamps
+	pullTimestamps := pullTimestamps(reader.lastPullTimestamp, reader.backFillEnabled)
+
+	var collectedMetrics []pdata.Metrics
+
+	// Pulling metrics for each generated pull timestamp
+	for _, pullTimestamp := range pullTimestamps {
+		stmtArgs := statementArgs{
+			query:                  reader.metricsMetadata.Query,
+			topMetricsQueryMaxRows: reader.topMetricsQueryMaxRows,
+			pullTimestamp:          pullTimestamp,
+		}
+
+		stmt := reader.statement(stmtArgs)
+		metrics, err := reader.pull(ctx, stmt)
+
+		if err != nil {
+			return nil, err
+		}
+
+		collectedMetrics = append(collectedMetrics, metrics...)
+	}
+
+	reader.lastPullTimestamp = pullTimestamps[len(pullTimestamps)-1]
+
+	return collectedMetrics, nil
+}
+
+// This slice will always contain at least one value.
+func pullTimestamps(lastPullTimestamp time.Time, backFillEnabled bool) []time.Time {
+	var timestamps []time.Time
+	upperBound := nowAtStartOfMinute()
+
+	if lastPullTimestamp.IsZero() {
+		if backFillEnabled {
+			timestamps = pullTimestampsWithMinuteDifference(upperBound.Add(-1*backFillIntervalDuration), upperBound)
+		} else {
+			timestamps = []time.Time{upperBound}
+		}
+	} else {
+		// lastPullTimestamp is already set to start of minute
+		timestamps = pullTimestampsWithMinuteDifference(lastPullTimestamp, upperBound)
+	}
+
+	return timestamps
+}
+
+// This slice will always contain at least one value.
+// Difference between each two points is 1 minute.
+func pullTimestampsWithMinuteDifference(lowerBound time.Time, upperBound time.Time) []time.Time {
+	var timestamps []time.Time
+
+	for value := lowerBound.Add(time.Minute); !value.After(upperBound); value = value.Add(time.Minute) {
```

Comment:
Do we backfill upto 59 minutes to avoid off by 1 issues from the line collector collects to the time prometheus server pulls the data.

Reply:
Yes, we sacrifice 1 minute due to processing lag (query execution, processing, etc.).